Enable reading validation via admin input

carlsonrob opened the pull request:
# Description
Currently, readings are not validated upon upload. This PR enables an admin to define default test values, which are then executed when readings are uploaded. The admin panel, meter modal (create/edit), meter object and database have been modified to accommodate these values. The conditionSet object was previously undefined, but can now be determined using these values and enable readings to be validated.
@TrongQuocLe @codehomie1

Fixes #511

## Type of change
- [ ] Note merging this changes the node modules
- [x] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist
- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request

## Limitations
Written test cases need to be created.

codehomie1 committed:
Merge remote-tracking branch 'spearec/reactstrap' into new_UI_setChecks

carlsonrob committed:
Set conditionSet through UI (Draft)

carlsonrob committed:
Update

carlsonrob committed:
Update

codehomie1 committed:
added minVal condset val

codehomie1 committed:
Merge branch 'new_UI_setChecks' of https://github.com/TrongQuocLe/OED into new_UI_setChecks

carlsonrob committed:
change conditionSetMap to conditionSet

codehomie1 committed:
add default condset vals to admin Panel

codehomie1 committed:
meter values default to admin values

carlsonrob committed:
Merge branch 'OpenEnergyDashboard:development' into new_UI_setChecks

Commit:
changed Meter Page UI

Commit:
resolve bugs

carlsonrob committed:
Include disableChecks in db

carlsonrob committed:
Add conditionSet objet to Mamac

carlsonrob committed:
Update Mamac w/ disabledChecks

carlsonrob committed:
Add conditionSet object to Obvius

carlsonrob committed:
Add conditionSet object to Egauge

Commit:
fix bugs

codehomie1 committed:
test preference api to save admin minval to db

codehomie1 committed:
fixed merge conflict in createmetermodalcomponent

Commit:
update UI

carlsonrob committed:
Add disableChecks to conditionSet for Mamac, Obvius and Egauge

carlsonrob committed:
Validate data if checks are enabled

carlsonrob committed:
Define conditionSet inline

carlsonrob committed:
Update meter field names

Commit:
default database for

Commit:
error to errors

Commit:
fix name values

codehomie1 committed:
made adminstate default same as db default

carlsonrob committed:
DB Update

Commit:
remove all disable checks in admin

Commit:
test

Commit:
test debug

Commit:
threshold to reading gap

Commit:
remove todo regarding add conditionSet params

codehomie1 committed:
change wording

codehomie1 committed:
update preferences stuff

carlsonrob committed:
Set default parameters for Meter

carlsonrob committed:
Modified default values

Commit:
added required, assigned moment for conditionSet

carlsonrob committed:
Update sql with default values

Commit:
added default conditionSet in Meter.js models

Commit:
trying to debug

Commit:
working admin.js

Commit:
added moment default in admin.js

carlsonrob committed:
Placeholder update

codehomie1 committed:
update minval and maxval values

codehomie1 committed:
update create meters minval minval to bigint

carlsonrob committed:
Update minVal, maxVal default parameter in Meter constructor

carlsonrob committed:
Update minVal, maxVal to bigInt in routes/meter

Commit:
added bound validation maxError, minVal, maxVal

Commit:
trying to git pull

Commit:
added minVal, maxVal, maxError bounds validation

Commit:
updated meter validation commennts

Commit:
added default conditionSet validation in admin pannel

Commit:
removed debug console.log

Commit:
removed used comment in routes/meters.js

Commit:
added a space removed on an import in PreferencesComponent.tsx

carlsonrob committed:
Reverse merging of tooltipmarkercomponent from eb3c78f74bc5e29a421b6caab3de616602ccd19c

carlsonrob committed:
Change minError to maxError in meter.js

codehomie1 committed:
update validate meter params type

codehomie1 committed:
db validation for min_val , max_val

codehomie1 committed:
fix type to bigint

carlsonrob committed:
Change minVal,maxVal to bigint to prevent error

Commit:
added maxDate and minDate validation in MeterModalComponent

Commit:
fix default reading value check type being string

Commit:
fixed code as suggested format

Commit:
try to resolve test

carlsonrob committed:
Real Precision Change

Commit:
remove unused comments

carlsonrob committed:
remove validIntervals as it's redundant in processData with msReadingGap comparision. modify test to use checkIntervals as validIntervals is removed

carlsonrob committed:
remove comment for validIntervals

Commit:
removed changes while testing

carlsonrob committed:
Remove space

Commit:
Added default meter min date, max date validation

huss committed:
Merge remote-tracking branch 'origin/development' into pr/carlsonrob/989

huss commented on src/server/models/Meter.js:
```diff
@@ -236,7 +251,7 @@ class Meter {
 		reading = this.reading, startTimestamp = this.startTimestamp, endTimestamp = this.endTimestamp,
 		previousEnd = this.previousEnd, unitId = this.unitId, defaultGraphicUnit = this.defaultGraphicUnit, areaUnit = this.areaUnit,
-		readingFrequency = this.readingFrequency) {
+		readingFrequency = this.readingFrequency, minVal = this.minVal, maxVal = this.maxVal, minDate = this.minDate, maxDate = this.maxDate, maxError = this.maxError, disableChecks = this.disableChecks) {
```
OED wants lines to be 150 characters or less. Please split this line.

huss commented on src/server/sql/preferences/insert_default_row.sql:
```diff
@@ -15,9 +15,10 @@ IF NOT EXISTS(SELECT *
+	default_area_normalization, default_area_unit, default_meter_reading_frequency, 
+	default_meter_minimum_value, default_meter_maximum_value, default_meter_minimum_date, 
+	default_meter_maximum_date, default_meter_reading_gap, default_meter_maximum_errors) 
+	VALUES ('', 'line', FALSE, 'en', NULL, 5, 25, FALSE, 'meters', '00:15:00', -9007199254740991, 9007199254740991, '1970-01-01 00:00:00+00:00', '6970-01-01 00:00:00+00:00', 0, 75);
```
OED wants lines to be 150 characters or less. Please split this line.

huss committed:
add migration

huss commented on src/client/app/actions/admin.ts:
```diff
@@ -104,13 +128,45 @@ function fetchPreferences(): Thunk {
+		moment(state.admin.defaultMeterMaximumDate) <= MAX_DATE_MOMENT &&
+		moment(state.admin.defaultMeterMinimumDate) <= moment(state.admin.defaultMeterMaximumDate) &&
+		(state.admin.defaultMeterMaximumErrors >= 0 && state.admin.defaultMeterMaximumErrors <= 75)
+	) return true;
```
OED likes to have  braces around if statement blocks. Also, in this case, I would use an else statement for the false case.

huss commented on src/client/app/actions/admin.ts:
```diff
@@ -104,13 +128,45 @@ function fetchPreferences(): Thunk {
 	};
 }
 
+/* Because there is no warn for invalid data in admin pannel, I use function validPreferences(state: State)
```
Could you make the comment about doing the tests here into a TODO so that it is done as other modal pages are when this page is changed? Also, pannel -> panel.

huss commented on src/client/app/actions/admin.ts:
```diff
@@ -104,13 +128,45 @@ function fetchPreferences(): Thunk {
 	return async (dispatch: Dispatch, getState: GetState) => {
 		const state = getState();
 		try {
+			if (!validPreferences(state)) throw new Error('invalid input');
```
if should have braces.

huss commented on src/client/app/components/admin/PreferencesComponent.tsx:
```diff
@@ -379,6 +470,37 @@ class PreferencesComponent extends React.Component<PreferencesPropsWithIntl> {
+	}
+
+	private handleDefaultMeterMaximumErrorsChange(e: { target: HTMLInputElement; }) {
+		this.props.updateDefaultMeterMaximumErrors(parseFloat(e.target.value));
```
Should this parse to an int? I know JS thinks about number but it would be a stronger test as the code will not really do a FP value.

huss commented on src/client/app/components/meters/CreateMeterModalComponent.tsx:
```diff
@@ -74,7 +75,13 @@ export default function CreateMeterModalComponent(props: CreateMeterModalCompone
+		minDate: adminState.defaultMeterMinimumDate,
+		maxDate: adminState.defaultMeterMaximumDate,
+		maxError: adminState.defaultMeterMaximumErrors,
+		disableChecks: false
```
I think this is a reasonable default. However, would it make sense to allow the admin input this value on the preferences page since they might not want tests on meters at their site? This would save changing for each new meter and also get automatically created meters to do the right thing if we use this value as is suggested elsewhere.

huss commented on src/client/app/components/meters/CreateMeterModalComponent.tsx:
```diff
@@ -126,8 +133,20 @@ export default function CreateMeterModalComponent(props: CreateMeterModalCompone
+	const MAX_VAL = Number.MAX_SAFE_INTEGER;
+	const MIN_DATE_MOMENT = moment(0).utc();
+	const MAX_DATE_MOMENT = moment(0).utc().add(5000, 'years');
+	const MIN_DATE_STRING = moment(0).utc().format('YYYY-MM-DD HH:mm:ssZ');
```
I know this does not change what the code does but I think it might be best to make the string values use the variables you just said above such as MIN_VAL. It will help if somebody changes the value in the future.

huss commented on src/client/app/components/meters/CreateMeterModalComponent.tsx:
```diff
@@ -138,7 +157,16 @@ export default function CreateMeterModalComponent(props: CreateMeterModalCompone
+			moment(state.minDate) >= MIN_DATE_MOMENT &&
+			moment(state.maxDate) <= MAX_DATE_MOMENT &&
+			moment(state.minDate) <= moment(state.maxDate) &&
+			(state.maxError >=0  && state.maxError <= 75)
```
You use the value 75 in a number of places so I wonder if it makes sense to create a const at the top for this value. Maybe export to use in other places?

huss commented on src/client/app/components/meters/CreateMeterModalComponent.tsx:
```diff
@@ -743,7 +824,54 @@ export default function CreateMeterModalComponent(props: CreateMeterModalCompone
+								onChange={e => handleStringChange(e)}
+								placeholder='YYYY-MM-DD HH:MM:SS'
+								required value={state.minDate}
+								invalid={state.minDate === '' || moment(state.minDate) < MIN_DATE_MOMENT || moment(state.minDate) > moment(state.maxDate)}/>
```
I'm surprised that this seems to work. moment has its own comparison functions so you should use something such as isBefore (https://momentjs.com/docs/#/query/is-same-or-before/).

I don't think we have done this anywhere else in the code but with the new type of checks it just hit me we might be  able to use the moment function isValid()  to verify that it is an okay date.  Right now I think it fails later in the process and is not as informative for the user as the new test on input. 

Also applies to max.

huss commented on src/client/app/components/meters/EditMeterModalComponent.tsx:
```diff
@@ -122,18 +128,50 @@ export default function EditMeterModalComponent(props: EditMeterModalComponentPr
+	const MAX_VAL = Number.MAX_SAFE_INTEGER;
+	const MIN_DATE_MOMENT = moment(0).utc();
+	const MAX_DATE_MOMENT = moment(0).utc().add(5000, 'years');
+	const MIN_DATE_STRING = moment(0).utc().format('YYYY-MM-DD HH:mm:ssZ');
```
A number of the comments for create also apply here and I will not repeat them.

huss commented on src/client/app/translations/data.js:
```diff
@@ -123,6 +123,12 @@ const localeData = {
 		"default.warning.file.size": "Default Warning File Size",
 		"default.file.size.limit": "Default File Size Limit",
 		"default.time.zone": "Default Time Zone",
+		"default.meter.minimum.value": "Default meter minimum reading value check",
```
The new translation keys are not in other languages.

huss commented on src/client/app/types/redux/admin.ts:
```diff
@@ -24,7 +24,13 @@ export type AdminAction =
+	| UpdateDefaultMeterMinimumDateAction
+	| UpdateDefaultMeterMaximumDateAction
+	| UpdateDefaultMeterMaximumErrorsAction
+	| UpdateDefaultMeterReadingGapAction;
```
A really small point but could MaximumErrors come last in the order as it is in other files/places.

huss commented on src/server/services/eGauge/readEgaugeData.js:
```diff
@@ -18,7 +18,7 @@ async function readEgaugeData(meter, conn) {
 	const meterReadings = await requestor.getMeterReadings();
 	await requestor.logout()
-
+	
```
Please remove the tab at the start of the line.

huss commented on src/server/services/obvius/loadLogfileToReadings.js:
```diff
@@ -52,7 +52,13 @@ async function loadLogfileToReadings(serialNumber, ipAddress, logfile, conn) {
 				undefined, // area unit
-				undefined // reading frequency
+				undefined, // reading frequency
+				undefined, // minVal
```
I wonder if we should use the site defaults here rather than undefined. They may often be the same due to the default values but they could differ. Note  would apply to the new reading gap in preferences too.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -627,7 +627,7 @@ async function processData(rows, meterID, timeSort = TimeSortTypesJS.increasing,
 	}
 	// Validate data if conditions given
-	if (conditionSet !== undefined && !validateReadings(result, conditionSet)) {
+	if (conditionSet !== undefined && !conditionSet['disableChecks'] && !validateReadings(result, conditionSet)) {
```
The processData.js uses appendMsgTotal() to add messages to what is returned to the user. I think it would be good to return these messages in the same way so that the caller receives the messages and not only goes to the log file. I know this is a non-trivial change so let us know if you prefer not to do it.

The error messages in validateReadings.js has the meter id. Could we pass the meter name as an argument so we could print that instead of the ID as is done in other messages? 

A couple of thoughts about the messages such as:
```js
log.warn(`ERROR WHEN CHECKING READING TIME FROM METER ${reading.meterID}: Time ${reading.startTimestamp} is earlier than lower bound ${minDate}`);
```
 First, OED does not use uppercase in messages. Second, I think this should be a log.error because it causes the readings to be rejected.

I think line 45 has startTimestamp when it should be endTimestamp so the code would be:
```js
log.warn(`ERROR WHEN CHECKING READING TIME FROM METER ${reading.meterID}: Time ${reading.endTimestamp} is later than upper bound ${maxDate}`);
```

I cannot put these comments in the other file so I'm putting it here.

huss commented on src/server/services/readMamacMeters.js:
```diff
@@ -96,7 +96,13 @@ async function getMeterInfo(url, ip, csvLine) {
 				undefined, // area unit
-				undefined // reading frequency
+				undefined, // reading frequency
+				undefined, // minVal
```
As suggested elsewhere, use the site defaults if that is done.

huss commented on src/server/sql/meter/create_meters_table.sql:
```diff
@@ -64,5 +64,11 @@ CREATE TABLE IF NOT EXISTS meters (
     area_unit area_unit_type NOT NULL DEFAULT 'none',
-    reading_frequency INTERVAL NOT NULL DEFAULT '00:15:00'
+    reading_frequency INTERVAL NOT NULL DEFAULT '00:15:00',
+    min_val DOUBLE PRECISION NOT NULL DEFAULT -9007199254740991 CHECK (min_val::DOUBLE PRECISION >= -9007199254740991),
```
The readings table uses FLOAT for the reading so maybe it would be best to use that here instead of double precision. Note all SQL changes need to be done similarly for the migrations.

huss commented on src/server/sql/preferences/create_preferences_table.sql:
```diff
@@ -14,5 +14,11 @@ CREATE TABLE IF NOT EXISTS preferences (
 	default_area_unit area_unit_type NOT NULL,
-	default_meter_reading_frequency INTERVAL NOT NULL
+	default_meter_reading_frequency INTERVAL NOT NULL,
+	default_meter_minimum_value DOUBLE PRECISION NOT NULL,
```
Change to float if done in meter.

huss commented:
To do testing, I merged in the latest development and also added the database migration that was needed. Since I had to do it I just pushed the changes myself. You can check if the migration seems correct to you and any changes in the PR SQL needs to stay in sync with this.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -627,7 +627,7 @@ async function processData(rows, meterID, timeSort = TimeSortTypesJS.increasing,
 	// Validate data if conditions given
-	if (conditionSet !== undefined && !validateReadings(result, conditionSet)) {
+	if (conditionSet !== undefined && !conditionSet['disableChecks'] && !validateReadings(result, conditionSet)) {
 		errMsg = `<h2>For meter ' + meterName + ': error when validating data so all reading are rejected</h2>`;
```
This is printing meterName. I think the code should be:
```js
`<h2>For meter ${meterName}: error when validating data so all reading are rejected</h2>`
```

huss reviewed (changes requested):
I want to thank @carlsonrob, @TrongQuocLe and @codehomie1 for their first contribution to OED. Overall, this is very good work and I found very few issues during testing. I have made a number of comments that should be addressed. Please let me know if you have any thoughts or questions.

Commit:
split line to meet 150-character limit

Commit:
add braces and TODO for preferences

Commit:
create and rename constants

carlsonrob committed:
Remove tab

carlsonrob committed:
Put UpdateDefaultMeterMaximumErrorsAction after UpdateDefaultMeterReadingGapAction

carlsonrob committed:
Print meterName variable

carlsonrob committed:
parseFloat to parseInt

Commit:
utilize moment functions for validation in Meter

Commit:
apply similar changes in Meter to preferences

TrongQuocLe committed:
Merge branch 'OpenEnergyDashboard:development' into new_UI_setChecks

Commit:
Add new translation keys in other languages.

Commit:
Change data type for min and max value in database

Commit:
Correct timestamp reference in error message

carlsonrob committed:
Integrate site preferences when instantiating a new Meter object

carlsonrob committed:
Add meterName parameter to validateReadings

Commit:
Add default disable checks

carlsonrob committed:
Append errMsg from validateReadings onto msgTotal in processData

carlsonrob committed:
Remove comment

Commit:
Add default disable checks

carlsonrob committed:
Fix TypeScript error by adding type guard for 'min' and 'max' properties

The TypeScript compiler was throwing errors because it couldn't find the properties 'min' and 'max' on the type 'BarReading'. The issue arises because TypeScript is unaware that 'min' and 'max' are only accessed when 'readings' is an array of 'LineReadings'. This commit adds a type guard to explicitly inform TypeScript about the type of 'reading', resolving the issue.

huss committed:
reuse value set in variable

Fixed for a comment that was not clear in my review to resolve this.

huss committed:
non-en has same order

huss committed:
Fix lint error

huss committed:
update migrations

huss committed:
add missing JSDoc param

huss committed:
blank line in validate html

huss commented:
I made a  few minor changes to the code. I don't think there are any issues but I'll give others one day to see if there are any issues before I merge this code.

huss reviewed (approved):
I want to thank all the developers of this code for being willing to implement additional features after the pull request was created, being so careful in finding a few things I missed and the speed with which they did all their work. The changes address all the comments and testing has found this to work well. This code is now ready to merge. Congratulations on your first contribution to OED.
